flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/sink/compactor/operator/CompactorOperatorStateHandler.java

```diff
@@ -141,26 +135,68 @@ public void processElement(
             throws Exception {
         Either<CommittableMessage<FileSinkCommittable>, CompactorRequest> record =
                 element.getValue();
-        if (stateDrained) {
+        if (passingThroughAll) {
```

Comment:
The implementation of this method looks a bit complex to me. From my view it would be better if we could keep a style of state machine: 

```
  Either<CommittableMessage<FileSinkCommittable>, CompactorRequest> record =
          element.getValue();
  if (passingThroughAll) {
      // all input should be committable messages to pass through
      output.collect(new StreamRecord<>(record.left()));
      return;
  } else if (passingThrough) {
      // 1. if it is a committable, emit it directly
      // 2. else if it is summary
      //  2.a if compacting result is empty
      //  2.b if not
  } else {
      // 1. if it is compacting request, submit it
      // 2. else if it is summary
      //  2.a if compacting result is empty
      //  2.b if not
      // 3. else if it is a committable
      //  3.a if it is not normal, deal with it
      //  3.b if it is normal, passing through = true
  }
```

We may change the `passingThrough` and `passingThroughAll` back to a state enumeration. 

Besides, we could extract the sub-process to handle the `CommittableSummary` and `normal / hiddle` Committables to private functions to further reduce the complexity of this method.

Reply:
I prefer not to make it a state enumeration since it's not so complex in my opinion. I suppose that a reorg of the code should be enough.